### ISSUE TITLE
numpy eigh not deterministic

### ISSUE THREAD

**Issue by DavidKershaw** [numpy/numpy#6378]
To illustrate the problem I need to attach a .py python script file and a .npz numpy binary savez file and github won't let me attach either of these file types. How can I do this?

**Comment by charris** [comment]
What do you mean by "not deterministic" ?.

**Comment by matthew-brett** [comment]
You could try uploading the files on dropbox or similar?

I seem to remember that routines like svd on Windows with ATLAS could give different answers when called twice on the same data (for example, flipped signs).

**Comment by seberg** [comment]
Frankly, I expect that is not a numpy issue. The underlying lapack implementation that is called may not guarantee the same results every time, not sure how the algorithms for eigh work, but different alignments, etc. I can imagine variations (or it might have bugs). That said, we should probably have a quick look to see that there is no bug in numpy.

**Comment by DavidKershaw** [comment]
Entire .py script appears below. The matrix ffm is very ill conditioned. The final print statement gives quite different answers depending on whether the first two print statements are commented out or not and with both print statements in it will randomly switch between the two different answers each time it is run.

import numpy as np
npzfile = np.load("D:\\almond\\dsk_research\\python\\bugtest.npz","rb")
# print npzfile.files
mmat=npzfile['mmat']
nmat=npzfile['nmat']
m=npzfile['m']
fr=np.linalg.solve(mmat,nmat)
ff=np.dot(nmat.transpose(),fr)
# print np.max(np.abs(mmat.transpose()-mmat))
ff=(ff+ff.transpose())/2.
ffm=m[:,np.newaxis]*ff*m
eigsys=np.linalg.eigh(ffm)
print eigsys[1][-2,-2], eigsys[1][-1,-2], eigsys[0][-2]

**Comment by DavidKershaw** [comment]
More precisely here is the behavior I get:
print 1 & print 2 both commented out - randomly switches between answer1 & answer2, mostly answer 2
print 1 in & print 2 commented out - randomly switches between answer1 & answer2, mostly answer 1
print 1 in or out & print 2 in - always get answer 1

**Comment by pv** [comment]
The cause most likely is that the data alignment decides which variant of sse based loops is run. This affects rounding errors, which are amplified by the bad conditioning. You can see if the results correlate with `your_array.__array_interface__['data'][0] % 16`. Modern compilers / blas libraries don't really guarantee floating-point determinism because this has a performance cost.

**Comment by DavidKershaw** [comment]
Thanks Pauli Virtanen for what I believe is the correct explanation. I am running the pythonxy installation of python 2.7.9 and numpy 1.8 on a 64bit windows 7 OS , Intel I5 cpu, box. Would this be using intel AVX or SSE or other earlier vectorization instructions? Would they all cause the same lack of floating-point determinism?

**Comment by charris** [comment]
@DavidKershaw Depends on whether 64 bit python is being used or not. You can check the size of `np.intp`
```
In [4]: np.intp(0).dtype.itemsize
Out[4]: 8
```
That is 64 bits for the pointer type on my 64 bit system.

**Comment by DavidKershaw** [comment]
On my computer I get:
>>> import numpy as np
>>> np.intp(0).dtype.itemsize
4
>>>

**Comment by charris** [comment]
Yep, 32 bit python. The 32 bit floating point has a lot of variations depending on which registers are used, storage of intermediate results, the SSE instruction set, etc. If that matters to you, a 64 bit system will be probably be more consistent.

**Comment by seberg** [comment]
OK, I am going to close the issue. Feel free to keep discussing.

**Comment by ysdtkm** [comment]
I encountered a similar problem with numpy.linalg.eigvals with a 64-bit system (RedHat Linux 6.9 with Anaconda3-5.1.0). I uploaded python script, binary of the matrix, error log, etc. to my repository.
https://github.com/ysdtkm/undeterministic_numpy

I couldn't reproduce the problem on my Windows Subsystems for Linux (Ubuntu 16.04 on Windows 10, with Anaconda3-5.1.0), and guess it is environment dependent.
